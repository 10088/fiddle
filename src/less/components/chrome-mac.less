.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 23px;
    font-size: 1.4rem;

    small,
    p,
    span {
      user-select: none;
    }
  }
}
