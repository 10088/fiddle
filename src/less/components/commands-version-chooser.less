button.version-chooser.bp3-button {
  border-radius: 3px;
}
