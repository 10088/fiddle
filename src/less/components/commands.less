@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-1;
  -webkit-app-region: drag;
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding: 10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      button {
        margin-left: 5px;
      }
    }
  }

  .address-bar {
    height: 28px;
    width: 360px;
    transition: width 0.4s;
    display: flex;

    input,
    span {
      font-family: @fonts-common;
      font-weight: 300;
      font-size: 11px;
      cursor: auto;
      -webkit-app-region: no-drag;
    }

    input {
      flex-grow: 1;
      background: white;
      padding-left: 1rem;
      height: 2.6rem;
    }

    input:invalid {
      box-shadow: inset 0 0 0 1px #e4252d;
    }

    span {
      background: white;
      color: #1d1d1d;
      border-bottom-color: rgb(209, 209, 209);
      border-bottom-style: solid;
      border-bottom-left-radius: 4px;
      border-bottom-width: 1px;
      border-image-outset: 0;
      border-image-repeat: stretch;
      border-image-slice: 100%;
      border-image-source: none;
      border-image-width: 1;
      border-left-color: rgb(209, 209, 209);
      border-left-style: solid;
      border-left-width: 1px;
      border-top-color: rgb(209, 209, 209);
      border-top-left-radius: 4px;
      border-top-style: solid;
      border-top-width: 1px;
      display: inline-block;
      height: 1.2rem;
      letter-spacing: normal;
      margin: 0;
      padding: 6px 0 10px 10px;
      user-select: text;
      width: 132px;
      text-rendering: auto;
      text-shadow: none;
      text-transform: none;
    }

    &.empty {
      width: 200px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }
}
