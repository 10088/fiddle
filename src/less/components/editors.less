.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 3px 0 29px;

    .mosaic-controls {
      transform: scale(0.7);
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
