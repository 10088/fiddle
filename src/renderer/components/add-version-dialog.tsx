import { observer } from 'mobx-react';
import * as path from 'path';
import * as React from 'react';
import * as semver from 'semver';

import { NpmVersion } from '../../interfaces';
import { getElectronNameForPlatform } from '../../utils/electron-name';
import { AppState } from '../state';
import { Dialog } from './dialog';

export interface AddVersionDialogProps {
  appState: AppState;
}

export interface AddVersionDialogState {
  isValidElectron: boolean;
  isValidVersion: boolean;
  file?: File;
  version: string;
}

/**
 * The "add version" dialog allows users to add custom builds of Electron.
 *
 * @class AddVersionDialog
 * @extends {React.Component<AddVersionDialogProps, AddVersionDialogState>}
 */
@observer
export class AddVersionDialog extends React.Component<AddVersionDialogProps, AddVersionDialogState> {
  constructor(props: AddVersionDialogProps) {
    super(props);

    this.state = {
      isValidVersion: false,
      isValidElectron: false,
      version: ''
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onClose = this.onClose.bind(this);
    this.onChangeFile = this.onChangeFile.bind(this);
    this.onChangeVersion = this.onChangeVersion.bind(this);
  }

  /**
   * Handles a change of the file input
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   */
  public async onChangeFile(event: React.ChangeEvent<HTMLInputElement>) {
    const { files } = event.target;
    const { binaryManager } = this.props.appState;
    const file = files && files[0] ? files[0] : undefined;

    const isValidElectron = !!(file
      && file.path
      && await binaryManager.getIsDownloaded('custom', file.path));

    this.setState({ file, isValidElectron });
  }

  public onChangeVersion(event: React.ChangeEvent<HTMLInputElement>) {
    const version = event.target.value || '';
    const isValidVersion = !!semver.valid(version);

    this.setState({
      version,
      isValidVersion,
     });
  }

  /**
   * Handles the submission of the dialog
   *
   * @returns {Promise<void>}
   */
  public async onSubmit(): Promise<void> {
    const { file, version } = this.state;

    if (!file) return;

    const name = file.path
      .slice(-20)
      .split(path.sep)
      .slice(1)
      .join(path.sep);

    const toAdd: NpmVersion = {
      localPath: file.path,
      version,
      name
    };

    this.props.appState.addLocalVersion(toAdd);
    this.onClose();
  }

  /**
   * Closes the dialog
   */
  public onClose() {
    this.props.appState.isAddVersionDialogShowing = false;
    this.reset();
  }

  /**
   * Reset this component's state
   */
  public reset(): void {
    this.setState({
      isValidElectron: false,
      isValidVersion: false,
      version: '',
      file: undefined
    });
  }

  get buttons() {
    const canSubmit = this.state.isValidElectron && this.state.isValidVersion;

    return [
      (
        <button
          className='button'
          key='submit'
          disabled={!canSubmit}
          onClick={this.onSubmit}
        >
          Add
        </button>
      ), (
        <button
          className='cancel'
          key='cancel'
          onClick={this.onClose}
        >
          Cancel
        </button>
      )
    ];
  }

  public render() {
    const { isAddVersionDialogShowing } = this.props.appState;
    const dirOptions = { webkitdirectory: 'true' };

    return (
      <Dialog
        isShowing={isAddVersionDialogShowing}
        isShowingBackdrop={true}
        buttons={this.buttons}
        onClose={this.onClose}
        isCentered={true}
        className='add-version-dialog'
        key='add-version-dialog'
      >
        <label
          htmlFor='custom-electron-version'
          className='force-button'
        >
          Select the folder containing {getElectronNameForPlatform()}
        </label>
        <input
          type='file'
          onChange={this.onChangeFile}
          id='custom-electron-version'
          name='custom-electron-version'
          {...dirOptions}
        />
        {this.renderPath()}
      </Dialog>
    );
  }

  private renderPath(): JSX.Element | null {
    const { file, isValidElectron } = this.state;

    if (!file || !file.path) return null;

    const info = isValidElectron
      ? `We found an ${getElectronNameForPlatform()} in this folder.`
      : `We did not find a ${getElectronNameForPlatform()} in this folder...`;

    return (
      <>
        <input
          readOnly={true}
          value={file.path}
        />
        <p>
          {info}
        </p>
        {this.renderVersionInput()}
      </>
    );
  }

  private renderVersionInput(): JSX.Element | null {
    const { isValidElectron, isValidVersion, version } = this.state;

    if (!isValidElectron) return null;

    return (
      <>
        <p>
          Please specify a version, used for typings and the name.
          Must be <code>semver</code> compliant.
        </p>
        <input
          className={isValidVersion ? '' : 'hasError'}
          value={version}
          onChange={this.onChangeVersion}
          placeholder='4.0.0'
        />
      </>
    );
  }

}
