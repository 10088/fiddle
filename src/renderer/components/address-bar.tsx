import * as React from 'react';
import { observer } from 'mobx-react';
import * as Octokit from '@octokit/rest';
import * as classNames from 'classnames';

import { AppState } from '../state';
import { INDEX_HTML_NAME, MAIN_JS_NAME, RENDERER_JS_NAME } from '../constants';
import { idFromUrl } from '../../utils/gist';
import { reaction } from 'mobx';

export interface AddressBarProps {
  appState: AppState;
}

export interface AddressBarState {
  value: string;
}

@observer
export class AddressBar extends React.Component<AddressBarProps, AddressBarState> {
  constructor(props: AddressBarProps) {
    super(props);

    this.loadFiddle = this.loadFiddle.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);

    this.state = {
      value: this.props.appState.gistId
    };
  }

  /**
   * Handle the form's submit event, trying to load whatever
   * URL was entered.
   *
   * @param {React.SyntheticEvent<HTMLFormElement>} event
   * @memberof AddressBar
   */
  public async handleSubmit(event: React.SyntheticEvent<HTMLFormElement>) {
    event.preventDefault();

    this.props.appState.gistId = idFromUrl(this.state.value) || this.state.value;
    if (this.state.value) {
      this.loadFiddle();
    }
  }

  /**
   * Once the component mounts, we'll subscribe to gistId changes
   */
  public componentDidMount() {
    reaction(
      () => this.props.appState.gistId,
      (gistId: string) => this.setState({ value: gistId })
    );
  }

  /**
   * Handle the change event, which usually just updates the address bar's value
   *
   * @param {React.ChangeEvent<HTMLInputElement>} event
   * @memberof AddressBar
   */
  public handleChange(event: React.ChangeEvent<HTMLInputElement>) {
    this.setState({ value: idFromUrl(event.target.value) || event.target.value });
  }

  /**
   * Load a fiddle
   *
   * @returns
   * @memberof AddressBar
   */
  public async loadFiddle() {
    const { gistId } = this.props.appState;

    if (!confirm('Are you sure you want to load a new fiddle, all current progress will be lost?')) return;

    try {
      const octo = new Octokit();
      const gist = await octo.gists.get({
        gist_id: gistId,
        id: gistId,
      });

      window.ElectronFiddle.app.setValues({
        html: gist.data.files[INDEX_HTML_NAME].content,
        main: gist.data.files[MAIN_JS_NAME].content,
        renderer: gist.data.files[RENDERER_JS_NAME].content,
      });
    } catch (error) {
      console.warn(`Loading fiddle failed`, error);
    }
  }

  public render() {
    const { isUnsaved } = this.props.appState;
    const { value } = this.state;
    const className = classNames('address-bar', isUnsaved, { empty: !value });

    return (
      <form className={className} onSubmit={this.handleSubmit}>
        <span>https://gist.github.com/</span>
        <input
          key='addressbar'
          type='text'
          placeholder='...'
          value={value}
          onChange={this.handleChange}
        />
      </form>
    );
  }
}
