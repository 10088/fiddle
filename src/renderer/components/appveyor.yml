environment:
  matrix:
    - nodejs_version: "8"

init:
  - git config --global core.symlinks true

install:
  - ps: Install-Product node $env:nodejs_version x64
  - node --version
  - npm i

cache:
  - '%APPDATA%\npm-cache -> appveyor.yml'

test_script:
  - npm run lint
  - npm run test

build: off
