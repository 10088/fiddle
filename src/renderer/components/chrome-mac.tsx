import { observer } from 'mobx-react';
import * as React from 'react';

import { getTitle } from '../../utils/get-title';
import { AppState } from '../state';

export interface ChromeMacProps {
  appState: AppState;
}

@observer
export class ChromeMac extends React.Component<ChromeMacProps> {
  public render() {
    if (process.platform !== 'darwin') return null;

    return (
      <div className='chrome drag'>
        <small>{getTitle(this.props.appState)}</small>
      </div>
    );
  }
}
