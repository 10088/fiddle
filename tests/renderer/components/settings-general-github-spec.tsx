import { shallow } from 'enzyme';
import * as React from 'react';

import { GitHubSettings } from '../../../src/renderer/components/settings-general-github';

describe('GitHubSettings component', () => {
  beforeEach(() => {
    this.store = {
      signOutGitHub: jest.fn()
    };
  });

  it('renders when not signed in', () => {
    const wrapper = shallow(
      <GitHubSettings appState={this.store} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders when signed in', () => {
    this.store.gitHubToken = '123';
    this.store.gitHubLogin = 'Test User';

    const wrapper = shallow(
      <GitHubSettings appState={this.store} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('handles a click on the signout button', () => {
    this.store.gitHubToken = '123';
    this.store.gitHubLogin = 'Test User';

    const wrapper = shallow(
      <GitHubSettings appState={this.store} />
    );

    wrapper.find('button').simulate('click');
    expect(this.store.signOutGitHub).toHaveBeenCalled();
  });
});
