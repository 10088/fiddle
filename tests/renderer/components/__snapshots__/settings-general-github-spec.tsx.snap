// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHubSettings component renders when not signed in 1`] = `
<div>
  <h4>
    GitHub
  </h4>
  <label
    key="signed-out-label"
  >
    We can publish your fiddles to GitHub as a Gist - that way you can share your fiddles with the world!
  </label>
  <button
    className="button"
    key="sign-in-button"
    onClick={[Function]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f2f6",
            "M416 448h-84c-6.6 0-12-5.4-12-12v-40c0-6.6 5.4-12 12-12h84c17.7 0 32-14.3 32-32V160c0-17.7-14.3-32-32-32h-84c-6.6 0-12-5.4-12-12V76c0-6.6 5.4-12 12-12h84c53 0 96 43 96 96v192c0 53-43 96-96 96zm-47-201L201 79c-15-15-41-4.5-41 17v96H24c-13.3 0-24 10.7-24 24v96c0 13.3 10.7 24 24 24h136v96c0 21.5 26 32 41 17l168-168c9.3-9.4 9.3-24.6 0-34z",
          ],
          "iconName": "sign-in-alt",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      transform={null}
    />
     Sign In
  </button>
</div>
`;

exports[`GitHubSettings component renders when signed in 1`] = `
<div>
  <h4>
    GitHub
  </h4>
  <label
    key="signed-in-label"
  >
    Using the personal access token you gave us, we logged you into GitHub as 
    Test User
    .
  </label>
  <button
    className="button"
    key="sign-out-button"
    onClick={[MockFunction]}
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f2f5",
            "M497 273L329 441c-15 15-41 4.5-41-17v-96H152c-13.3 0-24-10.7-24-24v-96c0-13.3 10.7-24 24-24h136V88c0-21.4 25.9-32 41-17l168 168c9.3 9.4 9.3 24.6 0 34zM192 436v-40c0-6.6-5.4-12-12-12H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h84c6.6 0 12-5.4 12-12V76c0-6.6-5.4-12-12-12H96c-53 0-96 43-96 96v192c0 53 43 96 96 96h84c6.6 0 12-5.4 12-12z",
          ],
          "iconName": "sign-out-alt",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      transform={null}
    />
     Remove Token & Sign Out
  </button>
</div>
`;
